core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java

```diff
@@ -1232,8 +1203,218 @@ public void run() {
+      {
+        transactionStates = new HashMap<Long, TransactionState>();
+        for (Long txid : txIdList)
+        {
+          //build ts object
+          TransactionState  ts = new TransactionState(txid);
+          ts.setDDLTx(true);
+          transactionStates.put(txid, ts);
+        }
+      }
+      return transactionStates;
+    }
+
+    private void recoverTransactions(Map<Long, TransactionState> transactionStates) throws IOException
+    {
+        if (LOG.isDebugEnabled()) LOG.debug("TRAF RCOV THREAD: in-doubt transaction size " + transactionStates.size());
+        
+        for (Map.Entry<Long, TransactionState> tsEntry : transactionStates.entrySet()) {
+            int isTransactionStillAlive = 0;
+            TransactionState ts = tsEntry.getValue();
+            Long txID = ts.getTransactionId();
+            // TransactionState ts = new TransactionState(txID);
+            
+            //It is possible for long prepare situations that involve multiple DDL
+            //operations, multiple prompts from RS is received. Hence check to see if there
+            //is a TS object in main TS list and transaction is still active.
+            //Note that tsEntry is local TS object. 
+            if (hbtx.mapTransactionStates.get(txID) != null) {
+              if (hbtx.mapTransactionStates.get(txID).getStatus().toString().contains("ACTIVE")) {
+                isTransactionStillAlive = 1;
+              }
+              if (LOG.isInfoEnabled()) 
+              LOG.info("TRAF RCOV THREAD: TID " + txID
+                        + " still has TS object in TM memory. TS details: "
+                        + hbtx.mapTransactionStates.get(txID).toString() 
+                        + " transactionAlive: " + isTransactionStillAlive);
+              if(isTransactionStillAlive == 1)
+                continue; //for loop
+            }
+           
+            try {
+                audit.getTransactionState(ts);
+                if (ts.getStatus().equals(TransState.STATE_COMMITTED.toString())) {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("TRAF RCOV THREAD:Redriving commit for " + txID + " number of regions " + ts.getParticipatingRegions().size() +
+                                " and tolerating UnknownTransactionExceptions");
+                    txnManager.doCommit(ts, true /*ignore UnknownTransactionException*/);
+                    if(useTlog && useForgotten) {
+                        long nextAsn = tLog.getNextAuditSeqNum((int)TransactionState.getNodeId(txID));
+                        tLog.putSingleRecord(txID, ts.getCommitId(), "FORGOTTEN", null, forceForgotten, nextAsn);
+                    }
+                } else if (ts.getStatus().equals(TransState.STATE_ABORTED.toString())) {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("TRAF RCOV THREAD:Redriving abort for " + txID);
+                    txnManager.abort(ts);
```

Comment:
@sbroeder :  Do we need true /*ignore UnknownTransactionException*/ for abort as well?